DESIGN.md

```diff
@@ -28,6 +28,196 @@ daemon is spawned to handle future requests in its stead.
 
 ![Arch](res/architecture.png)
 
+## Versioning
+
+This section describes S3 Server's support for the specific AWS S3 Bucket
+Versioning feature.
+
+### AWS S3 Bucket Versioning
+
+An object in a versioning enabled bucket can have multiple versions, whereas an
+object can have only a single version in a standard bucket without versioning.
+
+With the standard S3 API, any update (PUT, DELETE) to an object in a versioning
+enabled bucket would create a new version of that object. Read operations such
+as GET and LIST would see the latest version of the objects. The table below
+shows such example, a `PUT foo` generates a new version `v2` and keeps `v1`;
+GET and LIST return the latest version of `foo`, which is `v2` in this case.
+
+| key | version | value |
+|-----|:-------:|:-----:|
+| foo |   v2    |   A   |
+| foo |   v1    |   B   |
+
+To work with all existing versions of the objects, S3 provides an extended
+API which uses the same operations of the standard API, but with versioning
+information specified. For example, in order to get the version `v1` of `foo`,
+users have to specify that version in the query, such as
+`GET /foo?versionId=v1`.
+
+A bucket can be non-versioning or versioning. The version of the objects of a
+bucket before it was versioning-enabled are called the "null" version; these
+versions are kept when a bucket is versioning enabled and are identified with
+the version identifier as "null". Turning off the versioning of a bucket would
+make all operations target the "null" versions of the objects, but would not
+delete all existing versions.
```

Comment:
It's misleading, I guess, because that changing the versioning status doesn't actually change what version the operations targets. I tried to update to be a little more accurate.